### PR TITLE
PDF export

### PR DESCRIPTION
**Pull request by w8r**
fixes: https://linkurious.atlassian.net/browse/LKE-7363

 - migrated build and devserver to [Vite](https://vitejs.dev/) (the existing one wouldn't even build with jsPDF)
 - added PDF export with jsPDF and svg2pdf
 - added proper configuration error handling

pdf-export

### CODE REVIEW

**Comment by w8r** [comment]
@Linkurious/developers please review

File: web/index.html [hunk 1]
```diff
@@ -3,13 +3,12 @@
...
     <title>Image export plugin</title>
-    <link rel="stylesheet" href="index-bundle.css" />
     <link rel="stylesheet" href="assets/css/styles.css" />
     <link rel="icon" type="image/x-icon" href="assets/favicon.ico" />
   </head>
 
   <body>
     <div class="page-container" id="root"></div>
-    <script src="index-bundle.js"></script>
+    <script src="index.tsx" type="module"></script>
```
Comment (davidrapin) [inline web/index.html:12]: It would make sense to document supported browsers for the plugin in the readme.
From what is see, this forces at least:
- edge 16+
- firefox 60+
- chrome 61+

see https://caniuse.com/es6-module
any other restrictions that could add constraints to browser support that I'm missing?

**Review by davidrapin — COMMENTED** [review]
Great feature, customers will love it :) 
I added some comments & questions.

File: web/index.html (hunk 1 above)
Comment (w8r) [reply web/index.html:12]: It was messed up by the old build script, normally Vite takes care of it. Thanks for finding this problem! Now it should look like this in the final artefact:

```html
...
<link rel="icon" type="image/x-icon" href="./assets/favicon.7fb0b556.ico" />
    <script type="module" crossorigin src="./assets/index.e04eeab3.js"></script>
    <link rel="stylesheet" href="./assets/index.fb4e8328.css">
  </head>

  <body>
    <div class="page-container" id="root"></div>
    ...
```

File: Jenkinsfile [hunk 2]
```diff
@@ -10,6 +10,7 @@ nodeJob {
   ]
   runBenchTests = true
   runPreReleaseOnUpload = false
+  runDependencyVersionCheck = false
```
[the remark this answers is not in the document]
Comment (w8r) [reply Jenkinsfile:13]: I don't fully understand the feature, it was failing the builds before even starting and I don't know how is it supposed to depend on the latest version and the latest version of what.

File: src/components/ConfigurationError.tsx [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+  error: Error;
+}
+
+export const ConfigurationError: FC<ConfigurationErrorProps> = () => (
+  <Layout>
+    <Layout.Content>
+      <Result
+        status="error"
+        title="Configuration error"
+        subTitle="We are working on it"
```
[the remark this answers is not in the document]
Comment (w8r) [reply src/components/ConfigurationError.tsx:14]: Agree, but I don't know what to say here rather, maybe "LKE API failed to respond or is misconfigured"?

**Comment by enys** [comment]
The target branch should most likely be develop ?

File: Jenkinsfile (hunk 2 above)
[the remark this answers is not in the document]
Comment (enys) [reply Jenkinsfile]: this checks that scoped linkurious packages match the current project version. It indeed needs to be false

File: package.json [hunk 4]
```diff
@@ -20,14 +20,14 @@
   "author": "Andrea Fantaccione",
   "scripts": {
     "clean": "rm -rf dist *.tgz *.lke",
-    "copy-assets": "mkdir -p dist/public && cp -R web/assets dist/public && cp web/index.html dist/public",
+    "copy-assets": "mkdir -p dist/public && cp -R web/assets dist/public",
     "build:project": "npm-run-all build:manifest build:public build:backend",
     "build:manifest": "mkdir -p dist && envsub src/manifest.json dist/manifest.json",
-    "build:public": "esbuild web/index.tsx --bundle --minify --outfile=dist/public/index-bundle.js --target=es6 --define:IS_DEV=false",
+    "build:public": "vite build",
```
Comment (enys) [inline package.json:26]: I'm curious of why the change form esbuild to vite ?

File: web/index.html (hunk 1 above)
Comment (davidrapin) [reply web/index.html:12]: thank you, so we are still using the `type="module"` attribute, which means no edge 15 compatibility. Since this is a departure from what we officially support for LKE, I would still recommend that the plugin readme makes beowser compatibility explicit.

**Comment by enys** [comment]
to get an asset, usable in the server, you can add the ci:upload label when ready and then download it form nexus, or update the plugins.test.txt file on server repo. (@davidrapin per your question today)

File: src/components/ConfigurationError.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (davidrapin) [reply src/components/ConfigurationError.tsx:14]: I think "Error while trying to read the configuration, please reload the page." makes sense:
- it explains what happened
- gives you a best guess of how to fix (probable temporary network error, reloading will probably fix)

File: package.json [hunk 5]
```diff
@@ -106,23 +106,26 @@
...
     "@ant-design/colors": "6.0.0",
     "@ant-design/icons": "4.7.0",
     "@linkurious/ogma": "4.1.4",
     "@linkurious/ogma-export-stitch": "0.0.6",
-    "@linkurious/ogma-linkurious-parser": "3.1.11",
-    "@linkurious/ogma-mini-map": "0.0.27",
-    "@linkurious/rest-client": "3.1.11",
+    "@linkurious/ogma-linkurious-parser": "3.1.4",
+    "@linkurious/ogma-mini-map": "0.0.43",
+    "@linkurious/rest-client": "3.1.4",
```
[the remark this answers is not in the document]
Comment (davidrapin) [reply package.json:119]: we just need to merge recent changes from develop

File: src/components/ConfigurationError.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (w8r) [reply src/components/ConfigurationError.tsx:14]: done

File: web/index.html (hunk 1 above)
Comment (w8r) [reply web/index.html:12]: Added legacy plugin to Vite that is supposed to deal with it by adding trickery like 
```
<script type="module">try{import.meta.url;import("_").catch(()=>1);}catch(e){}window.__vite_is_modern_browser=true;</script>
```
and then automatic polyfills and fall back to the legacy build with no modules.

File: src/components/Preview/Modal.tsx [hunk 6]
```diff
@@ -58,7 +58,7 @@ export const Modal: FC<Props> = ({ visible, onCancel, onOk }) => {
     scalingStyleRule
       .destroy()
       .then(() =>
-        svg(ogma)
+        svg(ogma as unknown as Ogma)
```
[the remark this answers is not in the document]
Comment (w8r) [reply src/components/Preview/Modal.tsx:61]: for some reason `LKOgma` doesn't want to be cast to `Ogma`. I will check

File: src/components/Preview/Modal.tsx [hunk 7]
```diff
@@ -58,7 +58,7 @@ export const Modal: FC<Props> = ({ visible, onCancel, onOk }) => {
     scalingStyleRule
       .destroy()
       .then(() =>
```
[the remark this answers is not in the document]
Comment (w8r) [reply src/components/Preview/Modal.tsx:60]: react requires some hoop-jumping to use `async` in `useEffect` hooks. I tried to do it here, not sure if it improves the readability or makes it worse.

File: package.json [hunk 8]
```diff
@@ -75,13 +74,15 @@
     "@types/svgo": "2.6.3",
     "@typescript-eslint/eslint-plugin": "5.7.0",
     "@typescript-eslint/parser": "5.7.0",
+    "@vitejs/plugin-legacy": "^2.3.0",
```
Comment (davidrapin) [inline package.json:77]: some non-pined deps left

File: package.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (davidrapin) [reply package.json:121]: please update to the latest version of the LKE dependencies (3.1.12), this should also solve the casting LkOgma to Ogma, if you use the same Ogma version as ogma-linkurious-parser

File: web/index.html (hunk 1 above)
Comment (davidrapin) [reply web/index.html:12]: OK got it! thank you.

File: package.json (hunk 4 above)
Comment (enys) [reply package.json:26]: Great just making sure I hadn't missed the point. Thx for the details

File: package.json [hunk 9]
```diff
@@ -104,25 +105,29 @@
...
     "serve": "13.0.2",
     "sharp": "0.30.7",
+    "terser": "5.15.1",
     "ts-jest": "28.0.2",
     "ts-node": "10.8.1",
-    "typescript": "4.6.4"
+    "typescript": "4.6.4",
+    "vite": "3.1.8"
   },
   "dependencies": {
```
Comment (davidrapin) [inline package.json:114]: linkurious/rest-client & linkurious/ogma-linkurious-parser should be:
- in the same version
- pinned
- in 3.1.12 (latest LKE)

**Review by davidrapin — COMMENTED** [review]
@w8r let me know when we are good to merge so that I can involve Xavier for QA testing.

File: package.json (hunk 9 above)
Comment (w8r) [reply package.json:114]: done, put 3.1.12

File: package.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (w8r) [reply package.json]: done, didn't solve the typing problem, but it's not a big deal

File: src/utils/download.ts [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+    // @ts-ignore
+    const pageWidth = pdf.getPageWidth() - margin * 2;
+    // @ts-ignore
+    const pageHeight = pdf.getPageHeight() - margin * 2;
+
+    // vertical cursor
+    let y = margin;
+    // fitting ratio canvas to page
+    let ratio = 1 / Math.max(svgWidth / pageWidth, svgHeight / pageHeight);
+    // seems like big sizes breaks it
```
[the remark this answers is not in the document]
Comment (w8r) [reply src/utils/download.ts]: old comment, don't remember what it was about.

**Comment by w8r** [comment]
@davidrapin @XavierDeMontis it's ready for testing

**Comment by enys** [comment]
Well no one has approved the PR

**Comment by w8r** [comment]
so I re-requested the review

**Review by davidrapin — APPROVED** [review]
sorry, my bad, I though I had approved already

**Comment by w8r** [comment]
thanks!
